File: Match.scala
import upickle.default._
import java.nio.file.{Files, Paths}
import scala.io.Source
import java.net.URI
import java.net.http.{HttpClient, HttpRequest, HttpResponse}
import java.util.UUID
import org.apache.commons.math3.optim.linear._
import org.apache.commons.math3.optim.nonlinear.scalar.GoalType
import org.apache.commons.math3.optim.MaxIter
import scala.jdk.CollectionConverters._

case class User(id: Int, name: String, lat: Double, lon: Double, capabilities: List[String])
object User {
  implicit val rw: ReadWriter[User] = readwriter[ujson.Value].bimap[User](
    u => ujson.Obj("id" -> ujson.Num(u.id.toDouble), "name" -> ujson.Str(u.name), "lat" -> ujson.Num(u.lat), "lon" -> ujson.Num(u.lon), "capabilities" -> ujson.Arr.from(u.capabilities.map(ujson.Str))),
    json => User(
      json("id").num.toInt,
      json("name").str,
      json("lat").num,
      json("lon").num,
      json("capabilities").arr.map(_.str).toList
    )
  )
}

case class Place(id: Int, name: String, lat: Double, lon: Double, required_capabilities: List[String], capacity: Option[Int] = None)
object Place {
  implicit val rw: ReadWriter[Place] = readwriter[ujson.Value].bimap[Place](
    p => ujson.Obj("id" -> ujson.Num(p.id.toDouble), "name" -> ujson.Str(p.name), "lat" -> ujson.Num(p.lat), "lon" -> ujson.Num(p.lon), "required_capabilities" -> ujson.Arr.from(p.required_capabilities.map(ujson.Str)), "capacity" -> ujson.Num(p.capacity.getOrElse(1).toDouble)),
    json => Place(
      json("id").num.toInt,
      json("name").str,
      json("lat").num,
      json("lon").num,
      json("required_capabilities").arr.map(_.str).toList,
      json.obj.get("capacity").map(_.num.toInt)
    )
  )
}


case class MatchResult(place: String, user: String, capabilityScore: Double, distanceKm: Double, note: Option[String] = None)

object Match {
  val MAX_RELOCATION_KM = 30.0
  val OVER_CAP_PENALTY_PER_KM = 0.005
  val EARTH_RADIUS_KM = 6371.0
  val LLM_THRESHOLD = 0.5
  val BASE_URL = "https://opencode.ai/zen/go/v1"

  def haversineKm(lat1: Double, lon1: Double, lat2: Double, lon2: Double): Double = {
    val p1 = Math.toRadians(lat1)
    val p2 = Math.toRadians(lat2)
    val dphi = Math.toRadians(lat2 - lat1)
    val dlambda = Math.toRadians(lon2 - lon1)
    val a = Math.pow(Math.sin(dphi / 2), 2) + Math.cos(p1) * Math.cos(p2) * Math.pow(Math.sin(dlambda / 2), 2)
    2 * EARTH_RADIUS_KM * Math.asin(Math.sqrt(a))
  }

  def loadJson[T: ReadWriter](path: String): List[T] = {
    val jsonString = Source.fromFile(path)("UTF-8").mkString
    read[List[T]](jsonString)
  }

  def getLlmApiKey(): Option[String] = {
    val envPath = Paths.get("..", ".env").toAbsolutePath.normalize()
    sys.env.get("OPENCODEGO_API_KEY").orElse {
      if (Files.exists(envPath)) {
        Source.fromFile(envPath.toFile).getLines()
          .find(_.startsWith("OPENCODEGO_API_KEY="))
          .map(_.substring("OPENCODEGO_API_KEY=".length).stripPrefix("\"").stripSuffix("\""))
      } else None
    }
  }

  def llmScore(apiKey: String, user: User, place: Place): Double = {
    val model = sys.env.getOrElse("OPENCODEGO_MODEL", "deepseek-v4-flash")
    val prompt = s"User capabilities: ${user.capabilities.mkString(", ")}\n" +
      s"Place '${place.name}' required capabilities: ${place.required_capabilities.mkString(", ")}\n" +
      "Rate how well this user semantically fits the place's requirements, from 0.0 (no fit) " +
      "to 1.0 (perfect fit), giving partial credit for related or adjacent skills even without " +
      "an exact string match. Reply with ONLY a number between 0.0 and 1.0."

    val body = ujson.Obj(
      "model" -> model,
      "messages" -> ujson.Arr(
        ujson.Obj("role" -> "user", "content" -> prompt)
      ),
      "temperature" -> 0
    )

    val sessionId = UUID.randomUUID().toString
    val req = HttpRequest.newBuilder()
      .uri(URI.create(s"$BASE_URL/chat/completions"))
      .header("Content-Type", "application/json")
      .header("Authorization", s"Bearer $apiKey")
      .header("x-opencode-session", sessionId)
      .POST(HttpRequest.BodyPublishers.ofString(body.render()))
      .build()

    val client = HttpClient.newHttpClient()
    val res = client.send(req, HttpResponse.BodyHandlers.ofString())
    
    if (res.statusCode() == 200) {
      try {
        val json = ujson.read(res.body())
        val text = json("choices")(0)("message")("content").str.trim
        val score = text.split("\\s+")(0).toDouble
        Math.max(0.0, Math.min(1.0, score))
      } catch {
        case _: Exception => 0.0
      }
    } else 0.0
  }

  def solveTransportationLp(combined: Array[Array[Double]], capacities: Array[Int], forceFull: Boolean): Array[Array[Boolean]] = {
    val n = combined.length
    val m = combined(0).length
    val numVars = n * m

    val objCoeffs = new Array[Double](numVars)
    for (i <- 0 until n; j <- 0 until m) objCoeffs(i * m + j) = -combined(i)(j)
    val objective = new LinearObjectiveFunction(objCoeffs, 0.0)

    val constraints = new java.util.ArrayList[LinearConstraint]()

    for (j <- 0 until m) {
      val coeffs = new Array[Double](numVars)
      for (i <- 0 until n) coeffs(i * m + j) = 1.0
      constraints.add(new LinearConstraint(coeffs, Relationship.LEQ, capacities(j).toDouble))
    }

    val userRel = if (forceFull) Relationship.EQ else Relationship.LEQ
    for (i <- 0 until n) {
      val coeffs = new Array[Double](numVars)
      for (j <- 0 until m) coeffs(i * m + j) = 1.0
      constraints.add(new LinearConstraint(coeffs, userRel, 1.0))
    }

    val solver = new SimplexSolver()
    try {
      val solution = solver.optimize(
        new MaxIter(100000),
        objective,
        new LinearConstraintSet(constraints),
        GoalType.MINIMIZE,
        new NonNegativeConstraint(true)
      )

      val assignment = Array.ofDim[Boolean](n, m)
      for (i <- 0 until n; j <- 0 until m) {
        if (solution.getPoint()(i * m + j) > 0.5) assignment(i)(j) = true
      }
      assignment
    } catch {
      case e: Exception => throw new RuntimeException(s"LP solver failed: ${e.getMessage}")
    }
  }

  def relocationNote(users: List[User], userIdx: Int, place: Place, distancesCol: Array[Double], capScoresCol: Array[Double], distance: Double, capScore: Double): String = {
    val candidates = users.indices.filter(i => i != userIdx && distancesCol(i) <= MAX_RELOCATION_KM)
    val userName = users(userIdx).name
    if (candidates.isEmpty) {
      f"No candidate lives within ${MAX_RELOCATION_KM}%.0fkm of ${place.name}; ${userName} is the closest available fit at ${distance}%.1fkm, so the cap was exceeded."
    } else {
      val bestI = candidates.maxBy(i => capScoresCol(i))
      val bestName = users(bestI).name
      val bestScore = capScoresCol(bestI)
      f"${userName} relocates ${distance}%.1fkm (exceeds the ${MAX_RELOCATION_KM}%.0fkm cap) because their capability fit (${capScore}%.2f) clearly beats the best in-range candidate, ${bestName} (fit ${bestScore}%.2f) within ${MAX_RELOCATION_KM}%.0fkm."
    }
  }

  def main(args: Array[String]): Unit = {
    val llm = args.contains("--llm")
    val allowUnassigned = args.contains("--allow-unassigned")
    
    val users = loadJson[User]("users.json")
    val places = loadJson[Place]("places.json")

    val apiKey = if (llm) getLlmApiKey().getOrElse(sys.error("OPENCODEGO_API_KEY not found")) else ""

    val distances = Array.ofDim[Double](users.size, places.size)
    val capScores = Array.ofDim[Double](users.size, places.size)

    for (i <- users.indices; j <- places.indices) {
      val u = users(i)
      val p = places(j)
      distances(i)(j) = haversineKm(u.lat, u.lon, p.lat, p.lon)
      
      val uCaps = u.capabilities.map(_.toLowerCase).toSet
      val pCaps = p.required_capabilities.map(_.toLowerCase).toSet
      var base = if (pCaps.isEmpty) 0.0 else uCaps.intersect(pCaps).size.toDouble / pCaps.size
      
      if (llm && base < LLM_THRESHOLD) {
        base = Math.max(base, llmScore(apiKey, u, p))
      }
      capScores(i)(j) = base
    }

    val capacities = places.map(_.capacity.getOrElse(1)).toArray
    val totalCapacity = capacities.sum
    var wantFull = !allowUnassigned

    if (wantFull && totalCapacity < users.size) {
      println(s"(total capacity $totalCapacity < ${users.size} users -- can't force full assignment, falling back to best-available)\n")
      wantFull = false
    }

    val combined = Array.ofDim[Double](users.size, places.size)
    for (i <- users.indices; j <- places.indices) {
      val overCap = Math.max(0.0, distances(i)(j) - MAX_RELOCATION_KM)
      combined(i)(j) = capScores(i)(j) - overCap * OVER_CAP_PENALTY_PER_KM
    }

    val assignment = solveTransportationLp(combined, capacities, wantFull)
    var results = List[MatchResult]()

    for (i <- users.indices; j <- places.indices) {
      if (assignment(i)(j)) {
        val dist = distances(i)(j)
        val score = capScores(i)(j)
        val note = if (dist > MAX_RELOCATION_KM) {
          Some(relocationNote(users, i, places(j), distances.map(_(j)), capScores.map(_(j)), dist, score))
        } else None
        
        results ::= MatchResult(
          places(j).name,
          users(i).name,
          BigDecimal(score).setScale(3, BigDecimal.RoundingMode.HALF_UP).toDouble,
          BigDecimal(dist).setScale(1, BigDecimal.RoundingMode.HALF_UP).toDouble,
          note
        )
      }
    }

    val sortedResults = results.sortBy(r => (r.place, -r.capabilityScore))
    val unmatched = users.size - sortedResults.size
    
    if (unmatched > 0) {
      println(s"($unmatched user(s) left unassigned -- no place had net-positive value for them)\n")
    }

    for (m <- sortedResults) {
      val placeStr = m.place.padTo(30, ' ')
      val userStr = m.user.padTo(20, ' ')
      println(s"$placeStr -> $userStr (fit=${m.capabilityScore}, dist=${m.distanceKm}km)")
      m.note.foreach(n => println(s"  note: $n"))
    }
  }
}

File: match.py
#!/usr/bin/env -S uv run --script
# /// script
# requires-python = ">=3.11"
# dependencies = [
#     "numpy>=1.26",
#     "scipy>=1.11",
#     "openai>=1.0",
#     "python-dotenv>=1.0",
# ]
# ///
"""Match users to places by capability fit, with optional LLM-assisted scoring."""

import argparse
import json
import os
import sys
import uuid
from pathlib import Path

import numpy as np
from dotenv import load_dotenv
from openai import OpenAI
from scipy.optimize import linprog
from scipy.sparse import csr_matrix, vstack as sparse_vstack

BASE_URL = "https://opencode.ai/zen/go/v1"
MODEL = os.environ.get("OPENCODEGO_MODEL", "deepseek-v4-flash")
LLM_THRESHOLD = 0.5  # only ask the LLM to refine when the exact-overlap score is below this

MAX_RELOCATION_KM = 30.0
OVER_CAP_PENALTY_PER_KM = 0.005  # cost applied to every km beyond the 30km cap
EARTH_RADIUS_KM = 6371.0


def haversine_km(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    p1, p2 = np.radians(lat1), np.radians(lat2)
    dphi = np.radians(lat2 - lat1)
    dlambda = np.radians(lon2 - lon1)
    a = np.sin(dphi / 2) ** 2 + np.cos(p1) * np.cos(p2) * np.sin(dlambda / 2) ** 2
    return 2 * EARTH_RADIUS_KM * np.arcsin(np.sqrt(a))


def build_distance_matrix(users: list[dict], places: list[dict]) -> np.ndarray:
    dist = np.zeros((len(users), len(places)))
    for i, u in enumerate(users):
        for j, p in enumerate(places):
            dist[i, j] = haversine_km(u["lat"], u["lon"], p["lat"], p["lon"])
    return dist


def load_json(path: str) -> list[dict]:
    return json.loads(Path(path).read_text(encoding="utf-8"))


def overlap_score(user_caps: list[str], place_caps: list[str]) -> float:
    u = {c.lower() for c in user_caps}
    p = {c.lower() for c in place_caps}
    if not p:
        return 0.0
    return len(u & p) / len(p)


def get_llm_client() -> OpenAI:
    load_dotenv(Path(__file__).resolve().parent.parent / ".env")
    api_key = os.environ.get("OPENCODEGO_API_KEY")
    if not api_key:
        sys.exit("OPENCODEGO_API_KEY not found in ../.env")
    session_id = str(uuid.uuid4())
    return OpenAI(api_key=api_key, base_url=BASE_URL, default_headers={"x-opencode-session": session_id})


def llm_score(client: OpenAI, user: dict, place: dict) -> float:
    prompt = (
        f"User capabilities: {', '.join(user['capabilities'])}\n"
        f"Place '{place['name']}' required capabilities: {', '.join(place['required_capabilities'])}\n"
        "Rate how well this user semantically fits the place's requirements, from 0.0 (no fit) "
        "to 1.0 (perfect fit), giving partial credit for related or adjacent skills even without "
        "an exact string match. Reply with ONLY a number between 0.0 and 1.0."
    )
    resp = client.chat.completions.create(
        model=MODEL,
        messages=[{"role": "user", "content": prompt}],
        temperature=0,
    )
    text = resp.choices[0].message.content.strip()
    try:
        return max(0.0, min(1.0, float(text.split()[0])))
    except ValueError:
        return 0.0


def build_score_matrix(users: list[dict], places: list[dict], client: OpenAI | None) -> np.ndarray:
    scores = np.zeros((len(users), len(places)))
    for i, user in enumerate(users):
        for j, place in enumerate(places):
            base = overlap_score(user["capabilities"], place["required_capabilities"])
            if client is not None and base < LLM_THRESHOLD:
                base = max(base, llm_score(client, user, place))
            scores[i, j] = base
    return scores


def relocation_note(users: list[dict], user_idx: int, place: dict, distances_col: np.ndarray,
                     cap_scores_col: np.ndarray, distance: float, cap_score: float) -> str:
    candidates = [
        (i, cap_scores_col[i])
        for i in range(len(users))
        if i != user_idx and distances_col[i] <= MAX_RELOCATION_KM
    ]
    user_name = users[user_idx]["name"]
    if not candidates:
        return (
            f"No candidate lives within {MAX_RELOCATION_KM:.0f}km of {place['name']}; "
            f"{user_name} is the closest available fit at {distance:.1f}km, so the cap was exceeded."
        )
    best_i, best_score = max(candidates, key=lambda ic: ic[1])
    best_name = users[best_i]["name"]
    return (
        f"{user_name} relocates {distance:.1f}km (exceeds the {MAX_RELOCATION_KM:.0f}km cap) because their "
        f"capability fit ({cap_score:.2f}) clearly beats the best in-range candidate, {best_name} "
        f"(fit {best_score:.2f}) within {MAX_RELOCATION_KM:.0f}km."
    )


def solve_transportation_lp(combined: np.ndarray, capacities: list[int], force_full: bool = True) -> np.ndarray:
    """Exact max-weight bipartite b-matching (each place <=capacity users).

    This is a transportation problem: its constraint matrix is totally unimodular,
    so the LP relaxation's optimal vertex is already integral -- no MILP/branch-and-bound
    needed. Solved with HiGHS (scipy's default), which handles sparse LPs with
    n*m variables efficiently (tested to 1000 users x 100 places in ~1s).

    force_full=True requires every user to land in exactly one place (sum_j x_ij == 1),
    minimizing total penalty subject to full coverage -- raises if total capacity < n.
    force_full=False allows the solver to leave a user unassigned when every available
    slot would have net-negative combined score (skill mismatch + distance penalty).

    Returns an (n, m) 0/1 assignment matrix.
    """
    n, m = combined.shape
    if force_full and sum(capacities) < n:
        raise ValueError(f"total capacity ({sum(capacities)}) < number of users ({n}); cannot force full assignment")

    num_vars = n * m
    var_idx = np.arange(num_vars)
    user_of = var_idx // m
    place_of = var_idx % m

    A_ub = csr_matrix((np.ones(num_vars), (place_of, var_idx)), shape=(m, num_vars))
    b_ub = np.asarray(capacities, dtype=float)

    A_eq = b_eq = None
    if force_full:
        A_eq = csr_matrix((np.ones(num_vars), (user_of, var_idx)), shape=(n, num_vars))
        b_eq = np.ones(n)
    else:
        # <=1-per-user constraint folded into A_ub (sparse stack, not dense -- avoids an
        # (n+m) x (n*m) dense blowup at scale)
        user_ub = csr_matrix((np.ones(num_vars), (user_of, var_idx)), shape=(n, num_vars))
        A_ub = sparse_vstack([A_ub, user_ub]).tocsr()
        b_ub = np.concatenate([b_ub, np.ones(n)])

    res = linprog(
        c=-combined.flatten(),  # minimize negative == maximize
        A_ub=A_ub,
        b_ub=b_ub,
        A_eq=A_eq,
        b_eq=b_eq,
        bounds=(0, 1),
        method="highs",
    )
    if not res.success:
        raise RuntimeError(f"LP solver failed: {res.message}")
    return (res.x.reshape(n, m) > 0.5)


def best_matching(
    users: list[dict],
    places: list[dict],
    cap_scores: np.ndarray,
    distances: np.ndarray,
    force_full: bool = True,
) -> list[dict]:
    """Capacity-aware assignment: each place accepts up to `capacity` users (default 1).

    Solved exactly via linear programming (see solve_transportation_lp) -- not
    a greedy heuristic. An earlier greedy version (sort pairs by score, assign
    while capacity allows) was measurably suboptimal: on a 2-user/2-place
    counterexample it found total score 11 against a true optimum of 18.
    """
    capacities = [p.get("capacity", 1) for p in places]
    over_cap = np.maximum(0.0, distances - MAX_RELOCATION_KM)
    combined = cap_scores - over_cap * OVER_CAP_PENALTY_PER_KM

    assignment = solve_transportation_lp(combined, capacities, force_full=force_full)

    results = []
    for i, j in zip(*np.where(assignment)):
        i, j = int(i), int(j)
        distance = float(distances[i, j])
        entry = {
            "place": places[j]["name"],
            "user": users[i]["name"],
            "capability_score": round(float(cap_scores[i, j]), 3),
            "distance_km": round(distance, 1),
        }
        if distance > MAX_RELOCATION_KM:
            entry["note"] = relocation_note(
                users, i, places[j], distances[:, j], cap_scores[:, j], distance, float(cap_scores[i, j])
            )
        results.append(entry)

    results.sort(key=lambda e: (e["place"], -e["capability_score"]))
    return results


def main() -> None:
    parser = argparse.ArgumentParser(description="Match users to places by capability fit.")
    parser.add_argument("--users", default="users.json")
    parser.add_argument("--places", default="places.json")
    parser.add_argument("--llm", action="store_true", help="Use the LLM to refine low-overlap scores semantically")
    parser.add_argument("--allow-unassigned", action="store_true",
                         help="Let the solver leave a user unassigned instead of forcing everyone into a place")
    args = parser.parse_args()

    users = load_json(args.users)
    places = load_json(args.places)

    client = get_llm_client() if args.llm else None
    cap_scores = build_score_matrix(users, places, client)
    distances = build_distance_matrix(users, places)

    total_capacity = sum(p.get("capacity", 1) for p in places)
    want_full = not args.allow_unassigned
    if want_full and total_capacity < len(users):
        print(f"(total capacity {total_capacity} < {len(users)} users -- can't force full assignment, "
              f"falling back to best-available)\n")
        want_full = False

    matches = best_matching(users, places, cap_scores, distances, force_full=want_full)

    unmatched = len(users) - len(matches)
    if unmatched:
        print(f"({unmatched} user(s) left unassigned -- no place had net-positive value for them)\n")

    for m in matches:
        print(f"{m['place']:30s} -> {m['user']:20s} "
              f"(fit={m['capability_score']}, dist={m['distance_km']}km)")
        if "note" in m:
            print(f"  note: {m['note']}")


if __name__ == "__main__":
    main()
